### ISSUE TITLE
Incompatible Licenses Apache2.0 & GPLv2 (gatsby-plugin-sharp)

### ISSUE THREAD

**Issue by einSelbst** [gatsbyjs/gatsby#17584]
## Summary

Hello! 
Because reasons I had to make sure all licenses in a certain gatsby project we have play well together. Unfortunately there is an issue.

As explained here https://www.apache.org/licenses/GPL-compatibility.html 

> the FSF has never considered the Apache License to be compatible with GPL version 2

There are several apache2.0 licensed packages in a typical gatsby project, eg. starter has 42
and one library which is GPLv2 licensed, so this is where I see a problem.

This specific library is https://github.com/Iwasawafag/node-potrace
I opened an [issue](https://github.com/Iwasawafag/node-potrace/issues/7) and a [PR](https://github.com/Iwasawafag/node-potrace/pull/9) regarding this topic, no feedback so far.

node-potrace is a dependency of [gatsby-plugin-sharp](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-sharp/package.json). [Sharp](https://www.npmjs.com/package/sharp) itself is apache2.0 licensed so I think the incompatibility is already on that level. Gatsby-plugin-sharp is licensed as MIT, however I miss the knowledge to evaluate if this is ok.

## Relevant information

FWIW I used https://github.com/davglass/license-checker and run it on the gatsbyjs/gatsby-starter-default and printed the licenses.

Here is the summary: https://github.com/einSelbst/gatsby-starter-default/blob/master/licenses-summary.txt

Here are the details: https://github.com/einSelbst/gatsby-starter-default/blob/master/licenses-details.json

## Question

I would like to know if gatsby thinks this is a relevant issue. I would think it's also in gatsby's own interest to make sure the code which is distributed doesn't infringe licenses.

**Comment by einSelbst** [comment]
Thank you for looking into this @davidbailey00 !

**Comment by m-allanson** [comment]
It looks like the node-potrace authors are open to updating the license to GPL v3, which may resolve this issue? https://github.com/Iwasawafag/node-potrace/issues/7

**Comment by thinwhiteduke80** [comment]
as far as I understand, using GPLv3 still requires you to release as GPLv3
https://tldrlegal.com/license/gnu-general-public-license-v3-(gpl-3)

**Comment by thinwhiteduke80** [comment]
> It looks like the node-potrace authors are open to updating the license to GPL v3, which may resolve this issue? [Iwasawafag/node-potrace#7](https://github.com/Iwasawafag/node-potrace/issues/7)

as far as I understand, using GPLv3 still requires you to release as GPLv3
https://tldrlegal.com/license/gnu-general-public-license-v3-(gpl-3)

**Comment by sternetj** [comment]
What is the status of this issue? We have a solution utilizing contentful and gatsby but making our code open source is not an option so this is kind of a make or break issue for my organization.

**Comment by pieh** [comment]
@sternetj The status is that this issue need to be resolved. Because `node-potrace` can't really be relicensed to more permissive license even the maintainer would want to do so, I think the only solution for us is to make breaking change in `gatsby-plugin-sharp` and remove `node-potrace` usage from it (that would let it remain `MIT` licensed). And there would be another package that would utilize `node-potrace` (but that would need to be `GPL`), but that would be up to individual projects to decide wether they want to use it or not.

Because this is not done yet, there might be solutions to workaround `node-potrace` being in `gatsby-plugin-sharp` dependencies:

Using `yarn resolutions` (would require to use `yarn` for it) - you can force `potrace` to be resolved to something else than actual `potrace` - if you are not using `tracedSVG` feature (this is the feature that relies on `potrace`). Then you can supply mock library that just make sure that imports still work

Here's example: https://github.com/pieh/dont-use-potrace-example/commit/a05eb1530792815de7c0aea2d0e419166e44f3bc

I ran https://www.npmjs.com/package/license-checker on project afterwards and it doesn't contain the `GPL` licensed `potrace` - see https://github.com/pieh/dont-use-potrace-example/blob/master/licenses.txt output of it.

Interestingly there is one mention of GPL license still:
```
├─ node-forge@0.9.0
│  ├─ licenses: (BSD-3-Clause OR GPL-2.0)
│  ├─ repository: https://github.com/digitalbazaar/forge
│  ├─ publisher: Digital Bazaar, Inc.
│  ├─ email: support@digitalbazaar.com
│  ├─ url: http://digitalbazaar.com/
│  ├─ path: /Users/misiek/test/dont-use-potrace-example/node_modules/node-forge
│  └─ licenseFile: /Users/misiek/test/dont-use-potrace-example/node_modules/node-forge/LICENSE
```

`node-forge` is transitive dependency of `webpack-dev-server` and I didn't inspect what the `OR` in licenses really mean (and what impact it can have), but I also am not a lawyer so even if I did look it up, I wouldn't be able to give advices on that part

**Comment by pieh** [comment]
I think the way we could permamently solve it (still breaking change) would be to make `potrace` a peerDependency and let user opt into installing it and let `gatsby-plugin-sharp` check if it's available and conditionally enable `potrace` related functionality.

**Comment by dannyphillips** [comment]
Any updates on this?

**Comment by adamsiembida** [comment]
So does this basically mean that any website created with Gatsby must be licensed as GPL and provide their source? Also that any past website created using Gatsby must be relicensed as GPL and provide their source?

This seems to be a massive issue that should be much more clearly warned about, like 🚨🚨🚨🚨WARNING🚨🚨🚨🚨

**Comment by KyleAMathews** [comment]
@adamsiembida licenses of build tools don't apply to what they produce when you run them. So your proprietary code being built by open source code is still proprietary when it comes out. Most compliers are open source for example.

**Comment by adamsiembida** [comment]
> 
> 
> @adamsiembida licenses of build tools don't apply to what they produce when you run them. So your proprietary code being built by open source code is still proprietary when it comes out. Most compliers are open source for example.

My bad! I didn't check to see that the plugin was only a part of the build tools 😳 I was considering using Gatsby, but I already worry about licenses in large open source projects, with the deep dependency trees. I was doing some web searches to see if anyone else found any licensing issues and ran across this.

**Comment by lippytm** [comment]
So it's TIME THEN ? To Build a New and Better License? I don't know? Maybe a Few NFT? Crypto too? Big intersections? Networking? It's Just me the Dead IDIOT? I'm a Cosmetic Drunk and I look at a lot of road kill stuff.

**Comment by lippytm** [comment]
I don't know but it would seem you would want both, but both unlimited, and both limited, in, when, where, how, both Want & or Don't want! Comedians are the Most Dangerous Weapons in the Universe! Space Aliens Lives Matter! Don't Eat the Space Aliens! Santa Clause is Drunk! Please call Call Super Artificial Intelligence Comedians? And that NFT Crypto Autobot Automations Time Machines Reruns "Drunk Robot" moment? The Pink Floyd!

**Comment by einSelbst** [comment]
> @adamsiembida licenses of build tools don't apply to what they produce when you run them. So your proprietary code being built by open source code is still proprietary when it comes out. Most compliers are open source for example.

@KyleAMathews Does this mean we could close this issue?

**Comment by wahidshafique** [comment]
Is this issue still relevant given what @KyleAMathews mentioned re build tools?

**Comment by wardpeet** [comment]
https://github.com/gatsbyjs/gatsby/pull/35328 will fix this issue as we vendored potrace and upgraded to GNU3 which is compatible with MIT.

**Comment by wardpeet** [comment]
We've added GPLv3 support which is compatible with MIT

**Comment by tony** [comment]
I think there's a misunderstanding re: license compatibility

The issue is the other way around: Are gatsby-plugin-sharp's dependencies virally licensed or not?

gatsby-plugin-sharp would need to drop the GPL package or find a way to make it so normal codebases aren't running afoul of viral clauses.

**Comment by jonmadison-amzn** [comment]
@gatsbyjs/potrace page: https://www.npmjs.com/package/@gatsbyjs/potrace

states License is GPL-3.0

direct link to "repository" provided on that page, https://github.com/tooolbox/node-potrace

links to (presumably) a different package, with a license of GPL-2.0

https://github.com/tooolbox/node-potrace

**Comment by remigius42** [comment]
Hi, first of all thank you for all your work on this project and I am sorry if I am reanimating this issue, I just wanted to ask first before opening a possibly redundant issue:
Based on your comments @wardpeet and @tony within this issue combined with pull requests [#2](https://github.com/gatsbyjs/node-potrace/pull/2) and [#3](https://github.com/gatsbyjs/node-potrace/pull/3) I am not sure whether the issue has actually been resolved, because while the [license file is now LGPL-3.0](https://github.com/gatsbyjs/node-potrace/blob/main/LICENSE), the [package.json still mentions GPL-3.0 (on `main`)](https://github.com/gatsbyjs/node-potrace/blob/34a92f8d3fadd8f14df030b1c4085830e932ecdf/package.json#L22).

For me personally, it's not that much of an issue, because on the CI pipeline I can exclude the package from the license checks, now that I have verified the `LICENSE` file, but I guess it might trip others as well. Therefore my questions are:

- Do you agree that the `package.json` should be changed? Based on https://spdx.org/licenses/ I assume the value should be either `LGPL-3.0-only` or `LGPL-3.0-or-later`, as `LGPL-3.0` seems to be deprecated.
- If you think `package.json` should be updated, would you like to do it yourself to avoid a foreign commit, or should I fork and create a pull request?
- If I should created the pull request, do you want to reopen this issue or would you prefer a new one?

Thank you for your time and best regards
Andreas

**Comment by selinger** [comment]
I am the copyright holder of Potrace, and a copyright holder of node-potrace, which is a derivative work of Potrace. I state unambiguously that the software at  https://github.com/gatsbyjs/node-potrace is in violation of my copyright, since in a commit on 2022-03-29, the license has been changed from the GPL (GNU General Public License) to the LGPL (GNU Lesser General Public License), which is a different and incompatible license. 

The licensing terms of Potrace are clear: Potrace and its derivative works can only be distributed under the terms of the GPL. Anybody who distributes Potrace under the terms of the LGPL is in violation. 

I just contacted @wardpeet and asked for this to be fixed. But I am also stating it here for the public record, to give notice to anybody who might have redistributed this software under an incorrect license that it is unauthorized and illegal to do so.

Thanks, -- Peter Selinger
